### ISSUE TITLE
Decompressing deflate with zlib specific wrapper fails

### ISSUE THREAD

**Issue by DoumanAsh** [alexcrichton/flate2-rs#164]
I'm not sure if it is actually a problem, but during my testing I found that particular HTTP endpoint serves `deflate` compressed data that `flate2` is unable to decompress.

I stored raw compressed data in this example https://play.rust-lang.org/?gist=a1664d259fe2026cbddae3a42c5a5fab&version=stable&mode=debug&edition=2015

I found there is actually similar issue for .NET decoder which may be caused by the way python compresses its data.

>but the .NET world struggles mightily with DEFLATE content that includes header and trailer bits included by zlib. As a simple example, this will fail:

So in short I wonder if `flate2` here is fine?

**Comment by oyvindln** [comment]
You have to use `ZlibDecoder` for deflate data with the zlib wrapper, which is the case with HTTP deflate. I don't know why it's named deflate rather than zlib in the HTTP specification.

**Comment by DoumanAsh** [comment]
Hmmm now looking at mdn

>Using the zlib structure (defined in RFC 1950) with the deflate compression algorithm (defined in RFC 1951).

So it is actually not correct for me to use `DeflateDecoder` at all?

I'm sorry for unnecessary opening issue then

**Comment by DoumanAsh** [comment]
@oyvindln btw is there a way to determine whether zlib wrapper present so that I could choose correct decoder automatically?

**Comment by oyvindln** [comment]
I'm not sure if there is a good way to do it, unlike the [gzip](https://en.wikipedia.org/wiki/Gzip#File_format) wrapper format, the zlib format doesn't contain a magic number that identifies it. You would have to test whether there is a [valid](https://tools.ietf.org/html/rfc1950) zlib header, and if there isn't, try raw deflate.

**Comment by DoumanAsh** [comment]
I guess I jsut should prefer gzip as it is more reliable
